compiler/rustc_session/src/lint/builtin.rs

```diff
@@ -2197,6 +2190,83 @@ declare_lint! {
     report_in_external_macro
 }
 
+declare_lint! {
+    /// The `pointer_structural_match` lint detects pointers used in patterns that do not
+    /// behave deterministically across optimizations.
+    ///
+    /// ### Example
+    ///
+    /// ```rust,compile_fail
+    /// #![deny(pointer_structural_match)]
+    /// fn foo(a: usize, b: usize) -> usize { a + b }
+    /// const FOO: fn(usize, usize) -> usize = foo;
+    /// fn main() {
+    ///     match FOO {
+    ///         FOO => {},
+    ///         _ => {},
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Previous versions of Rust allowed function pointers and wide raw pointers in patterns.
+    /// While these work in many cases as expected by users, it is possible that due to
+    /// optimizations pointers are "not equal to themselves" or pointers to different functions
+    /// compare as equal during runtime. This is because LLVM optimizations can deduplicate
+    /// functions if their bodies are the same, thus also making pointers to these functions point
+    /// to the same location. Additionally functions may get duplicated if they are instantiated
+    /// in different crates and not deduplicated again via LTO.
+    pub POINTER_STRUCTURAL_MATCH,
+    Allow,
+    "pointers are not structural-match",
+    @future_incompatible = FutureIncompatibleInfo {
+        reference: "issue #62411 <https://github.com/rust-lang/rust/issues/70861>",
+        edition: None,
+    };
```

Comment:
Do we have any reason to ever phase this out? If it's not relevant to const generics, pattern-matching should be fine.

Reply:
We disallow floating points in patterns because they have "weird" equality. Function pointers have "even weirder" equality, so IMO we should either disallow both or allow both.

Also see https://github.com/rust-lang/rust/issues/70861.

Reply:
Function pointers don't have values where `x == x` returns `false`, floats do.
Maybe we should've never banned floats either, hmpf, at least as long as they're not NaNs.

Reply:
Function pointers have cases where `x == y` is true today but false after recompilation. IMO that's worse.